include/open62541pp/services/detail/ClientService.h

```diff
@@ -0,0 +1,114 @@
+#pragma once
+
+#include <cassert>
+#include <tuple>
+#include <type_traits>
+#include <utility>  // exchange, forward, move
+
+#include "open62541pp/Client.h"
+#include "open62541pp/ErrorHandling.h"
+#include "open62541pp/TypeRegistry.h"
+#include "open62541pp/async.h"
+#include "open62541pp/detail/Result.h"
+#include "open62541pp/detail/ScopeExit.h"
+#include "open62541pp/open62541.h"
+
+namespace opcua::services::detail {
+
+struct MoveResponse {
+    template <typename Response>
+    [[nodiscard]] constexpr auto operator()(Response&& value) noexcept {
+        return std::exchange(std::forward<Response>(value), {});
+    }
+};
+
+template <typename Request, typename Response, typename TransformResponse, typename CompletionToken>
+auto sendRequest(
+    Client& client,
+    const Request& request,
+    TransformResponse&& transformResponse,
+    CompletionToken&& token
+) {
+    static_assert(std::is_invocable_v<TransformResponse, Response&>);
+    using Result = std::invoke_result_t<TransformResponse, Response&>;
+
+    auto init = [&](auto completionHandler, auto transform) {
+        using CompletionHandler = decltype(completionHandler);
+        using Context = std::tuple<TransformResponse, CompletionHandler>;
+
+        auto callback = [](UA_Client*, void* userdata, uint32_t /* reqId */, void* responsePtr) {
+            assert(userdata != nullptr);
+            std::unique_ptr<Context> context{static_cast<Context*>(userdata)};
+            auto& handler = std::get<CompletionHandler>(*context);
+
+            try {
+                if (responsePtr == nullptr) {
+                    if constexpr (std::is_void_v<Result>) {
+                        std::invoke(handler, UA_STATUSCODE_BADUNEXPECTEDERROR);
+                    } else {
+                        Result emptyResult{};
+                        std::invoke(handler, UA_STATUSCODE_BADUNEXPECTEDERROR, emptyResult);
+                    }
+                }
+
+                Response& response = *static_cast<Response*>(responsePtr);
+                auto result = opcua::detail::tryInvoke(
+                    std::get<TransformResponse>(*context), response
+                );
+                if constexpr (std::is_void_v<Result>) {
+                    std::invoke(handler, result.code());
+                } else {
+                    std::invoke(handler, result.code(), *result);
+                }
+            } catch (const std::exception&) {
```

Comment:
This catch would currently swallow uncaught std exceptions from the completion handler, for example a std::out_of_range etc., right?

Wouldn't it be better to let the exception be uncaught and let the application potentially crash, in case the completion handler doesn't take care of it?
If there is a scenario where such exceptions need to be caught outside of the completion handler, a user could invoke client.run() within a try-catch block.

Reply:
I would prefer not to catch the exceptions like Asio is doing it. My biggest concern is, that open62541 is not exception safe and an exception might break a control flow and lead to memory leaks!?

Maybe we just have test this with sanitizers :)

Reply:
That's a good point and I think it might be safer to set a constraint that the completion handler is `noexcept` then? 
From user perspective I would find that a reasonable constraint, seeing that open62541pp is a wrapper around a C lib :)

Reply:
I totally agree. I see two options:
1. document this
2. enforce noexcept with e.g. a `static_assert(std::is_nothrow_invocable<...>)`

The second option would expect every lambda and function passed as a callback to be explicitly marked as `noexcept`. This is not really practicable so this makes me favor the first option.

Reply:
Both approaches sound okay to me, I have some questions:

1. Is there a way to safely prevent potential memory leaks in this case, or would you document it sth like "programmer's risk" if they try to catch the exception outside of the callback?
2. Why do you think marking the callbacks explicitly with except is not practicable? For me it would not be practicable if that forbid me to call throwing functions within the callback, but that part still works

Reply:
1. An option could be an general-purpose exception handler for both the server and client to allow customization of error handling.
    ```cpp
    using ExceptionHandler = std::function<void(std::exception_ptr)>;
    Server::setExceptionHandler(ExceptionHandler handler);
    ```
    The user decides:
    1. log the exceptions
    2. do nothing
    3. rethrow the exception and catch it around e.g. `Server::run()`. This may cause memory leaks (I have to double check the event loop implementation of open62541)

2. I personally wouldn't like to always specify the noexcept especially for simple lambdas `[](StatusCode code) noexcept {}`... But I could get used to it 😀

Reply:
Your suspicion was right, I just tried it out with a modified client method async example and I did get a bunch of memory leak detections by the sanitizer :-|

I usually try to make it as difficult as possible for a user to stumble into potential pitfalls, that's why I'm a fan of preventing situations ii. and iii., and even no. i. can quickly be missed when an exception is only logged but not actually handled by the user.
Also with a centralized exception handler defined by the library it becomes more difficult to decide (and for the user to understand) when an exception should be caught by the central handler and when it should be forwarded to the user. One could name it something like a "centralized callback exception handler", though. But maybe still difficult to grasp 🤔 

With a callback-specific approach, maybe it helps to provide a default logging wrapper or swallow wrapper for the callable for a user to re-use? But I think that wouldn't help your original intention of not having to specify something extra 😄

Reply:
Thanks for testing! Sadly though...

The exception handler approach disassociates the exception from its source. I guess this would be ok if exceptions are only indicate exceptional conditions, e.g. `std::bad_alloc`. One popular HTTP client library is using a similar approach to approach exceptions in user-defined callbacks: https://github.com/yhirose/cpp-httplib?tab=readme-ov-file#exception-handler

Another idea could be the extension of the `Result` type to allow a result and either a status code or an exception pointer. This is similar to Boost's Outcome design of [outcome::result<T, EC>](https://ned14.github.io/outcome/tutorial/essential/result/) (with error code) and [`outcome::outcome<T, EC, EP>`](https://ned14.github.io/outcome/tutorial/essential/outcome/) (with error code or exception pointer).
I really like binding the exception pointer to the result, but this results in an overhead of +8 bytes.

Reply:
Is there a a distinction between exceptions such as std::bad_alloc and user-defined exceptions? I used to work more without exceptions in the past, so I'm still quite new to that field :-)

I don't have experience working with HTTP, but looking at the protocol I think there are differences in a HTTP scenario compared to OPC UA: HTTP is stateless, so if I understood it right, then from a server perspective an exception handler has the complete context information if it also receives the initial request, as is the case with the handler in cpp-httplib. In comparison OPC UA is stateful, which means even a hypothetical inclusion of the initial request in an exception handler could not serve as full context information.

I guess 8 Bytes, which only add up to the stack in most situations, wouldn't make a perceivable difference. Though I'm not sure if I'm following your idea about the Result type right 🤔 
If you extend the Result type to hold an exception, that exception could only be coming from the TransformResponse handler but not the CompletionHandler, right?
The TransformResponse handler looks like something internal to open62541pp and I don't see how it could throw an exception, because you are only wrapping already-allocated data from the C stack I think? The completion handler running the user code on the other hand would still throw as before.

Reply:
Yes I agree, it's tedious to include exceptions in the Result only for such very rare edge cases :(

If you want to be on the safe side and consider exceptions such as `std::bad_alloc` in the TransformResponse, what about catching the exception and representing it as `UA_STATUSCODE_BADOUTOFMEMORY`? That status code seems to be only used in a server context (as per internet opinion), but open62541 already uses it on the client side in a couple of places.
Although I have no idea how much of an edge case `std::bad_alloc` is anyway - I can hardly imagine a system that runs with dynamic heap allocations and then would not voluntarily or involuntarily crash/restart anyway when it runs out of memory.

Reply:
Yes good idea, just map `std::bad_alloc` to `UA_STATUSCODE_BADOUTOFMEMORY` 👍 

So let's try to enforce the user callbacks to be noexcept. We can later add simple wrapper to make throwing callbacks into noexcept logging callbacks as you suggested, e.g.:

```cpp
template <typename F>
class LogException {
public:
    explicit LogException(F&& func)
        : func_(std::forward<F>(func)) {}

    template <typename... Args>
    auto operator()(Args&&... args) noexcept {
        try {
            return func_(std::forward<Args>(args)...);
        } catch (const std::exception& e) {
            std::cout << e.what() << std::endl;
        }
    }

private:
    F func_;  
};

// noexcept callback
services::readNodeClassAsync(client, id, [](StatusCode code, NodeClass nodeClass) noexcept {
    // ...
});

// wrapped callback
services::readNodeClassAsync(client, id, LogException{[](StatusCode code, NodeClass nodeClass) {
    // ...
}});
```

Reply:
The `ExceptionCatcher` is now integrated and allows us the Asio-like approach of propagating the exceptions to the `run`/`runIterate` methods: https://github.com/open62541pp/open62541pp/pull/111/commits/6f684c005bfac902dc08fffda25632acf737b27b

I added tests to check the behavior of exceptions thrown in both the transform function and the user-defined callback: https://github.com/open62541pp/open62541pp/blob/6f684c005bfac902dc08fffda25632acf737b27b/tests/ClientService.cpp